File: python/keymaster/src/keymaster/net.py
"""Guards for fetches to a caller-supplied host.

Remote name lookup takes a domain from the caller and fetches
``https://<domain>/.well-known/names...``, which makes it an SSRF primitive
unless the target is checked: the classic payload is a cloud metadata address
such as ``169.254.169.254``, reachable from inside a container and happy to
hand out credentials (#252).

Mirrors ``packages/common/src/net.ts`` and its Node transport. Both ports served this surface with
no target check at all on the public paths, so they are fixed together.
"""

from __future__ import annotations

import asyncio
import ipaddress
import socket
from typing import Any, Optional

import httpx

MAX_REDIRECTS = 3

# The statuses that are actually redirects. Treating the whole 3xx range as one
# turns a 304 Not Modified -- which carries no Location -- into a "redirect with
# no location" error, where httpx would simply have returned the response.
REDIRECT_STATUSES = frozenset({301, 302, 303, 307, 308})

# Names that resolve inside a host or a private network by convention. This
# cannot catch a public name pointed at a private address -- that is DNS
# rebinding, and it needs resolve-then-pin rather than a string check.
BLOCKED_SUFFIXES = (".localhost", ".local", ".internal", ".home.arpa")


def _parse_ipv4_any(hostname: str) -> Optional[int]:
    """Parse every form ``inet_aton`` accepts, as an integer.

    ``ipaddress`` deliberately rejects these, but the platform resolver does
    not, so ``2130706433``, ``0177.0.0.1``, ``0x7f000001`` and ``127.1`` all
    reach 127.0.0.1 regardless of what ``ipaddress`` thinks of them.
    """
    parts = hostname.split(".")

    if not parts or len(parts) > 4:
        return None

    values: list[int] = []

    for part in parts:
        if not part:
            return None

        try:
            if part.lower().startswith("0x"):
                value = int(part, 16)
            elif part.startswith("0") and len(part) > 1:
                value = int(part, 8)
            else:
                value = int(part, 10)
        except ValueError:
            return None

        if value < 0:
            return None

        values.append(value)

    # With fewer than four parts the last one covers the remaining bytes:
    # 127.1 is 127.0.0.1, and a lone 2130706433 is the whole address.
    last = values.pop()
    remaining = 4 - len(values)

    if last >= 256 ** remaining:
        return None

    if any(value > 255 for value in values):
        return None

    address = last

    for index, value in enumerate(values):
        address += value * (256 ** (3 - index))

    return address if address <= 0xFFFFFFFF else None


# Ranges the ipaddress flags do not cover. Carrier-grade NAT is not marked
# private by is_private, but a name resolving into it is still pointing at
# infrastructure rather than at a host the caller meant to reach, and the
# TypeScript port blocks it -- the two must agree or the guard depends on
# which language a caller happens to be using.
_EXTRA_BLOCKED = (ipaddress.ip_network("100.64.0.0/10"),)


def _is_blocked_address(address: Any) -> bool:
    # An IPv4 address wearing an IPv6 hat is judged on the IPv4 it carries.
    mapped = getattr(address, "ipv4_mapped", None)
    if mapped is not None:
        address = mapped

    if any(address in network for network in _EXTRA_BLOCKED if network.version == address.version):
        return True

    return bool(
        address.is_private
        or address.is_loopback
        or address.is_link_local
        or address.is_multicast
        or address.is_reserved
        or address.is_unspecified
    )


def is_private_hostname(hostname: str) -> bool:
    if not isinstance(hostname, str):
        return True

    host = hostname.strip().lower().rstrip(".")

    if not host or host == "localhost":
        return True

    # A URL keeps IPv6 literals in brackets; a bare address may arrive without.
    candidate = host[1:-1] if host.startswith("[") and host.endswith("]") else host

    # Zone identifiers (fe80::1%eth0) are not part of the address.
    if "%" in candidate:
        candidate = candidate.split("%", 1)[0]

    if ":" in candidate:
        try:
            return _is_blocked_address(ipaddress.IPv6Address(candidate))
        except ValueError:
            return True  # looks like an IPv6 literal but is not one

    numeric = _parse_ipv4_any(candidate)
    if numeric is not None:
        return _is_blocked_address(ipaddress.IPv4Address(numeric))

    return any(host.endswith(suffix) for suffix in BLOCKED_SUFFIXES)


async def resolve_public_addresses(host: str, port: int, timeout: float) -> list[str]:
    # getaddrinfo follows CNAMEs and returns final IPv4/IPv6 destinations. Reject
    # mixed public/private answers rather than relying on address-selection order.
    answers = await asyncio.wait_for(
        asyncio.get_running_loop().getaddrinfo(host, port, type=socket.SOCK_STREAM),
        timeout=timeout,
    )
    addresses = list(dict.fromkeys(answer[4][0] for answer in answers))
    if not addresses or any(is_private_hostname(address) for address in addresses):
        raise ValueError(f"refusing DNS resolution to private address for {host}")
    return addresses


async def fetch_public_https(
    method: str,
    url: str,
    headers: dict[str, str] | None = None,
    json_body: Any | None = None,
    timeout: float = 30.0,
) -> httpx.Response:
    """Validate every hop and connect to a validated IP, retaining Host/TLS identity."""
    current = url
    for _ in range(MAX_REDIRECTS + 1):
        parsed = httpx.URL(current)
        if parsed.scheme != "https":
            raise ValueError(f"refusing non-https request to {parsed.host}")
        if is_private_hostname(parsed.host):
            raise ValueError(f"refusing request to private address {parsed.host}")

        deadline = asyncio.get_running_loop().time() + timeout
        addresses = await resolve_public_addresses(parsed.host, parsed.port or 443, timeout)
        request_headers = httpx.Headers(headers)
        request_headers["Host"] = httpx.Request(method, parsed).headers["Host"]
        # A fresh client prevents pooling TLS sessions/cookies by the substituted
        # IP across distinct hostnames. Proxies must not re-resolve the hostname.
        async with httpx.AsyncClient(timeout=timeout, follow_redirects=False, trust_env=False) as client:
            for index, address in enumerate(addresses):
                try:
                    remaining = deadline - asyncio.get_running_loop().time()
                    if remaining <= 0:
                        raise httpx.ConnectTimeout("public HTTPS connection timeout")
                    response = await client.request(
                        method, parsed.copy_with(host=address), headers=request_headers, json=json_body,
                        timeout=remaining,
                        # Public httpcore extension keeps certificate verification
                        # and SNI against the original hostname, not the pinned IP.
                        # https://www.encode.io/httpcore/extensions/#sni_hostname
                        extensions={"sni_hostname": parsed.host},
                    )
                    break
                except (httpx.ConnectError, httpx.ConnectTimeout):
                    if index == len(addresses) - 1:
                        raise
        if response.status_code not in REDIRECT_STATUSES:
            return response
        location = response.headers.get("location")
        if not location:
            raise ValueError(f"redirect with no location from {parsed.host}")
        current = str(parsed.join(location))

    raise ValueError(f"too many redirects ({MAX_REDIRECTS})")

File: packages/keymaster/src/net.ts
// Guards for fetches to a caller-supplied host.
//
// Remote name lookup takes a domain from the caller and fetches
// https://<domain>/.well-known/names..., which makes it an SSRF primitive
// unless the target is checked: the classic payload is a cloud metadata
// address such as 169.254.169.254, reachable from inside a container and
// happy to hand out credentials (#252).
//
// This runs in browser wallets as well as Node, so `node:net` is not
// available and the address parsing here is done by hand.

// Every way inet_aton will accept an IPv4 address. `fetch` hands the hostname
// to the platform resolver, which accepts all of these, so a check that only
// understands dotted-decimal is a check that can be walked around:
//
//   2130706433   0177.0.0.1   0x7f000001   127.1
//
// all reach 127.0.0.1. Returns the address as a 32-bit number, or null when
// the string is not an IPv4 address in any of those forms.
function parseIpv4(hostname: string): number | null {
    const parts = hostname.split('.');

    if (parts.length === 0 || parts.length > 4) {
        return null;
    }

    const values: number[] = [];

    for (const part of parts) {
        if (part === '') {
            return null;
        }

        let value: number;

        if (/^0[xX][0-9a-fA-F]+$/.test(part)) {
            value = parseInt(part, 16);
        }
        else if (/^0[0-7]+$/.test(part)) {
            value = parseInt(part, 8);
        }
        else if (/^[0-9]+$/.test(part)) {
            value = parseInt(part, 10);
        }
        else {
            return null;
        }

        if (!Number.isFinite(value) || value < 0) {
            return null;
        }

        values.push(value);
    }

    // With fewer than four parts the last one covers the remaining bytes:
    // 127.1 is 127.0.0.1, and a lone 2130706433 is the whole address.
    const last = values.pop()!;
    const remainingBytes = 4 - values.length;

    if (last >= Math.pow(256, remainingBytes)) {
        return null;
    }

    if (values.some(value => value > 255)) {
        return null;
    }

    let address = last;

    for (let i = 0; i < values.length; i++) {
        address += values[i] * Math.pow(256, 3 - i);
    }

    return address >>> 0;
}

// Ranges that should never be the target of a lookup on behalf of a caller.
// Loopback and the RFC 1918 blocks are the obvious ones; 169.254/16 is the
// link-local range that carries cloud metadata, and the rest are reserved
// space that a public name has no business resolving to.
const BLOCKED_IPV4: Array<[string, number]> = [
    ['0.0.0.0', 8],        // "this host"
    ['10.0.0.0', 8],       // RFC 1918
    ['100.64.0.0', 10],    // carrier-grade NAT
    ['127.0.0.0', 8],      // loopback
    ['169.254.0.0', 16],   // link-local, incl. cloud metadata
    ['172.16.0.0', 12],    // RFC 1918
    ['192.0.0.0', 24],     // IETF protocol assignments
    ['192.168.0.0', 16],   // RFC 1918
    ['198.18.0.0', 15],    // benchmarking
    ['224.0.0.0', 4],      // multicast
    ['240.0.0.0', 4],      // reserved, incl. broadcast
];

function inRange(address: number, network: string, bits: number): boolean {
    const base = parseIpv4(network)!;
    // A /0 would shift by 32, which JavaScript treats as a shift by 0.
    const mask = bits === 0 ? 0 : (0xffffffff << (32 - bits)) >>> 0;
    return (address & mask) >>> 0 === (base & mask) >>> 0;
}

function isPrivateIpv4(address: number): boolean {
    return BLOCKED_IPV4.some(([network, bits]) => inRange(address, network, bits));
}

// Expands an IPv6 address to its sixteen bytes, or null if it is not one.
// Written out rather than delegated because there is no parser available in
// both runtimes this package targets.
function parseIpv6(hostname: string): number[] | null {
    let text = hostname;

    // A URL keeps IPv6 literals in brackets; a bare address may arrive without.
    if (text.startsWith('[') && text.endsWith(']')) {
        text = text.slice(1, -1);
    }

    // Zone identifiers (fe80::1%eth0) are not part of the address.
    const zone = text.indexOf('%');
    if (zone !== -1) {
        text = text.slice(0, zone);
    }

    if (!text.includes(':')) {
        return null;
    }

    const halves = text.split('::');
    if (halves.length > 2) {
        return null;
    }

    function toGroups(section: string): number[][] | null {
        if (section === '') {
            return [];
        }

        const groups: number[][] = [];

        for (const piece of section.split(':')) {
            // A trailing IPv4 part, as in ::ffff:127.0.0.1, is four more bytes.
            if (piece.includes('.')) {
                const address = parseIpv4(piece);
                if (address === null) {
                    return null;
                }
                groups.push([(address >>> 24) & 0xff, (address >>> 16) & 0xff]);
                groups.push([(address >>> 8) & 0xff, address & 0xff]);
                continue;
            }

            if (!/^[0-9a-fA-F]{1,4}$/.test(piece)) {
                return null;
            }

            const value = parseInt(piece, 16);
            groups.push([(value >> 8) & 0xff, value & 0xff]);
        }

        return groups;
    }

    const head = toGroups(halves[0]);
    const tail = halves.length === 2 ? toGroups(halves[1]) : [];

    if (head === null || tail === null) {
        return null;
    }

    const missing = 8 - head.length - tail.length;

    if (halves.length === 2) {
        if (missing < 0) {
            return null;
        }
    }
    else if (missing !== 0) {
        return null;
    }

    const filled = [...head, ...Array.from({ length: Math.max(missing, 0) }, () => [0, 0]), ...tail];

    return filled.flat();
}

function isPrivateIpv6(bytes: number[]): boolean {
    const isZero = (from: number, to: number) => bytes.slice(from, to).every(byte => byte === 0);

    // ::1 loopback, and :: unspecified.
    if (isZero(0, 15) && (bytes[15] === 1 || bytes[15] === 0)) {
        return true;
    }

    // ::ffff:0:0/96 -- an IPv4 address wearing an IPv6 hat, so judge the IPv4.
    if (isZero(0, 10) && bytes[10] === 0xff && bytes[11] === 0xff) {
        const embedded = ((bytes[12] << 24) | (bytes[13] << 16) | (bytes[14] << 8) | bytes[15]) >>> 0;
        return isPrivateIpv4(embedded);
    }

    // fc00::/7 unique local, fe80::/10 link-local.
    if ((bytes[0] & 0xfe) === 0xfc) {
        return true;
    }

    return (bytes[0] === 0xfe && (bytes[1] & 0xc0) === 0x80);
}

// Names that resolve inside a host or a private network by convention. This
// cannot catch a public name pointed at a private address -- that is DNS
// rebinding, and it needs resolve-then-pin rather than a string check -- but
// it does catch the names that are private by definition.
const BLOCKED_SUFFIXES = ['.localhost', '.local', '.internal', '.home.arpa'];

export function isPrivateHostname(hostname: string): boolean {
    if (typeof hostname !== 'string' || hostname === '') {
        return true;
    }

    const host = hostname.trim().toLowerCase().replace(/\.$/, '');

    if (host === '' || host === 'localhost') {
        return true;
    }

    const ipv6 = parseIpv6(host);
    if (ipv6) {
        return isPrivateIpv6(ipv6);
    }

    const ipv4 = parseIpv4(host);
    if (ipv4 !== null) {
        return isPrivateIpv4(ipv4);
    }

    return BLOCKED_SUFFIXES.some(suffix => host.endsWith(suffix));
}

const MAX_REDIRECTS = 3;

// Fetches over https, refusing any hop that is not https or that points at a
// private target. Checking only the first URL is not enough: `fetch` follows
// redirects on its own, so a public host answering 302 with a Location of
// http://169.254.169.254/ reaches the address the check exists to keep out.
// The Lightning LUD-16 path already walks redirects this way for the scheme;
// this adds the target check that path does not have.
export async function fetchPublicHttps(target: string, init?: RequestInit): Promise<Response> {
    let current = target;

    for (let hop = 0; hop <= MAX_REDIRECTS; hop++) {
        const url = new URL(current);

        if (url.protocol !== 'https:') {
            throw new Error(`refusing non-https request to ${url.host}`);
        }

        if (isPrivateHostname(url.hostname)) {
            throw new Error(`refusing request to private address ${url.hostname}`);
        }

        const response = await fetch(current, { ...init, redirect: 'manual' });

        // Follow only what is definitely a redirect. Testing for the negative
        // instead would treat an absent or non-numeric status as one and go
        // looking for a Location header that was never there.
        if (!(response.status >= 300 && response.status < 400)) {
            return response;
        }

        const location = response.headers.get('location');

        if (!location) {
            throw new Error(`redirect with no location from ${url.host}`);
        }

        current = new URL(location, current).toString();
    }

    throw new Error(`too many redirects (${MAX_REDIRECTS})`);
}
